doc/dvsource-file.1
.\" dvsource-file.1 written by Ben Hutchings <ben@decadent.org.uk>
.TH DVSOURCE-FILE 1 "18 February 2009"
.SH NAME
dvsource-file \- file source for DVswitch
.SH SYNOPSIS
.HP
.B dvsource-file
.RB [ \-h
.IR host ]
.RB [ \-p
.IR port ]
.RB [ \-l ]
.I filename
.SH DESCRIPTION
.LP
Stream a DV file to the mixer.  By default this plays the file once
and then exits.  This is mostly useful for development and testing
purposes.
.SH OPTIONS
.TP
.BI \-h " host " \-p " port"
.RS
Specify the network address on which DVswitch is listening.  The host
address may be specified by name or as an IPv4 or IPv6 literal.
.RE
.TP
.B \-l
.RS
Play the file repeatedly in a loop.
.RE
.SH AUTHOR
Ben Hutchings <ben@decadent.org.uk>.
.SH SEE ALSO
/usr/share/doc/dvswitch/README


ChangeLog
dvswitch (0.8) unstable; urgency=low

  * Added ticks and labels to the VU-meter
  * Bug fixes:
    - Removed use of function not present in Gtkmm 2.8
    - Fixed warnings from CMake 2.6
    - Fixed warnings from GCC
    - Fixed drag-selection when the pointer moves beyond the left or top
      edge of the display
    - Fixed memory corruption/crash in a selection corner-case
    - Fixed synchronisation of display buffers with the X server
    - Fixed repainting of main display before any frames have been received

 -- Ben Hutchings <ben@decadent.org.uk>  Wed, 04 Feb 2009 23:16:13 +0000

dvswitch (0.8~rc1) unstable; urgency=low

  * Disabled recording at startup
  * Added menu, currently with Quit command
  * Added buttons to GUI for effect selection
  * Changed key commands slightly
  * Allowed changing of primary and secondary video sources without
    cancelling the effect
  * Made dvsink-files create directories as necessary
  * Made dvsource-file reject non-DV files
  * Added support for USB DV sources through dvgrab and the uvcvideo driver
  * Moved source selector buttons to keep window within 1024x768 (depending
    on font selection and window decoration)
  * Added VU-meter for basic audio monitoring
  * Added warning indicator for sources which don't match the selected video
    and audio format
  * Documentation changes:
    - Documented all key commands
    - Changed contact address to dvswitch-devel@lists.alioth.debian.org
    - Added notes to (maybe) avoid user mistakes
  * Build system changes:
    - Changed to use cmake
    - Fixed building with current ffmpeg libavcodec library
  * Bug fixes:
    - Changed picture-in-picture selection to maintain frame aspect ratio
    - Restricted picture-in-picture to use the active area of the reduced
      frame
    - Slightly improved colour accuracy of picture-in-picture
    - Fixed cancellation of picture-in-picture selection
    - Fixed crash when user tries to apply picture-in-picture with an empty
      selection
    - Fixed crash when display depth is not 24 or 32
    - Fixed crash when users click in the dvswitch window without any source
      enabled
    - Made dvsource-file work with non-seekable files such as pipes

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 27 Jan 2009 01:04:43 +0000

dvswitch (0.7) unstable; urgency=low

  * Made keyboard shortcut for cut button work
  * Moved video/DV system parameters into a common structure
  * Replaced libdv codec with DV codec in libavcodec (ffmpeg), which
    is significantly faster
  * Added UI for picture-in-picture mixing

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 20 Apr 2008 22:18:37 +0100

dvswitch (0.6.2) unstable; urgency=low

  * Fixed bug in thumbnail image size initialisation that could result
    in a fatal BadValue error from X
  * Fixed minor bug in thumbnail image cleanup
  * Fixed bug that caused all sinks to be disabled when not recording

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 24 Feb 2008 10:55:41 +0100

dvswitch (0.6.1) unstable; urgency=low

  * Improved X driver compatibility:
    - Fixed calculation of Xv port numbers
    - Changed selection of visual class and depth for images to match the
      window
    - Added error messages and later checks for failure of image creation,
      where previously dvswitch could crash

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 21 Feb 2008 22:17:32 +0000

dvswitch (0.6) unstable; urgency=low

  * Added some infrastructure for video effects including
    picture-in-picture mixing
  * Switched from gtkmm 2.0/2.2 to 2.4+
  * Corrected small error in video display width
  * Changed the server to indicate overflow in output to sinks rather
    than disconnecting sinks that have overflowed; the file sink will
    create a new file after each overflow
  * Implemented dynamic display sizing, adding support for 16:9 frames
  * Turned source selection images into radio buttons
  * Fixed labelling of sources numbered above 9
  * Added support for starting and stopping recording while leaving
    any stream running
  * Added control buttons for recording and changed to default
    mnemonic for Cut

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 19 Feb 2008 09:37:22 +0000

dvswitch (0.5.1) unstable; urgency=low

  * Replaced FIREWIRE_DEVICE and --device option to dvsource-firewire
    with FIREWIRE_CARD and --card option
  * Corrected arguments to dvgrab

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 10 Jun 2007 21:41:17 +0100

dvswitch (0.5.0) unstable; urgency=low

  * Added indicator of title-safe area

 -- Ben Hutchings <ben@decadent.org.uk>  Mon, 30 Apr 2007 02:23:08 +0100

dvswitch (0.4) unstable; urgency=low

  * Implemented a command-spawning sink
  * Improved error handling
  * Added ability to switch between audio sources
  * Added indicators of selected video and audio sources

 -- Ben Hutchings <ben@decadent.org.uk>  Sun, 15 Apr 2007 16:27:47 +0100

dvswitch (0.3) unstable; urgency=low

  * Implemented dubbing of audio source onto video source (and
    silencing if audio source is unavailable)
  * Implemented reuse of source slots, so a disconnected source that
    reconnects will normally reappear in the same place
  * Added support for digit keys on the numeric keypad
  * Fixed priority of display updates, so the user interfaceshould
    remain responsive even with silly numbers of sources
  * Implemented synchronisation of mixer clock with audio source clock
  * Implemented sink serving and a DIF file sink
  * Fixed various bugs

 -- Ben Hutchings <ben@decadent.org.uk>  Tue, 10 Apr 2007 01:54:32 +0100

dvswitch (0.2) unstable; urgency=low

  * Fixed hang on exit
  * Implemented thumbnail displays of all sources
  * Switched to 1-based source selection in order to match keyboard layout
  * Reduced display lag
  * Reduced processor load of display updates by using Xv and Xshm

 -- Ben Hutchings <ben@decadent.org.uk>  Fri,  6 Apr 2007 00:54:22 +0100

dvswitch (0.1) unstable; urgency=low

  * Initial release


src/dvsource-file.c
/* Copyright 2007-2009 Ben Hutchings.
 * Copyright 2008 Petter Reinholdtsen.
 * See the file "COPYING" for licence details.
 */

#include <assert.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <fcntl.h>
#include <getopt.h>
#include <sys/types.h>
#include <unistd.h>
#include <netinet/in.h>

#include "config.h"
#include "dif.h"
#include "frame_timer.h"
#include "protocol.h"
#include "socket.h"

static struct option options[] = {
    {"host",   1, NULL, 'h'},
    {"port",   1, NULL, 'p'},
    {"loop",   0, NULL, 'l'},
    {"help",   0, NULL, 'H'},
    {NULL,     0, NULL, 0}
};

static char * mixer_host = NULL;
static char * mixer_port = NULL;

static void handle_config(const char * name, const char * value)
{
    if (strcmp(name, "MIXER_HOST") == 0)
    {
	free(mixer_host);
	mixer_host = strdup(value);
    }
    else if (strcmp(name, "MIXER_PORT") == 0)
    {
	free(mixer_port);
	mixer_port = strdup(value);
    }
}

static void usage(const char * progname)
{
    fprintf(stderr,
	    "\
Usage: %s [{-h|--host} MIXER-HOST] [{-p|--port} MIXER-PORT] [-l] FILE\n",
	    progname);
}

struct transfer_params {
    int            file;
    int            sock;
    bool           opt_loop;
};

static ssize_t read_retry(int fd, void * buf, size_t count)
{
    ssize_t chunk, total = 0;

    do
    {
	chunk = read(fd, buf, count);
	if (chunk <= 0)
	{
	    if (total == 0)
		return chunk;
	    break;
	}
	total += chunk;
	buf = (char *)buf + chunk;
	count -= chunk;
    }
    while (count);

    return total;
}

static void transfer_frames(struct transfer_params * params)
{
    const struct dv_system * last_system = 0, * system;
    static uint8_t buf[DIF_MAX_FRAME_SIZE];
    uint64_t frame_timestamp = 0;
    unsigned int frame_interval = 0;

    frame_timer_init();

    for (;;)
    {
	ssize_t size = read_retry(params->file, buf, DIF_SEQUENCE_SIZE);
	if (size == 0)
	{
	    // End of file; exit or loop
	    if (!params->opt_loop)
		return;
	    if (lseek(params->file, 0, 0) == 0)
		continue;
	    perror("ERROR: lseek");
	    exit(1);
	}
	if (size != (ssize_t)DIF_SEQUENCE_SIZE)
	{
	    if (size < 0)
		perror("ERROR: read");
	    else
		fputs("ERROR: Failed to read complete frame\n", stderr);
	    exit(1);
	}

	system = dv_buffer_system(buf);

	/* (Re)set the timer according to this frame's video system. */
	if (system != last_system)
	{
	    last_system = system;
	    frame_timestamp = frame_timer_get();
	    frame_interval = (1000000000 / system->frame_rate_numer
			      * system->frame_rate_denom);
	}

	size = read_retry(params->file, buf + DIF_SEQUENCE_SIZE,
			  system->size - DIF_SEQUENCE_SIZE);
	if (size != (ssize_t)(system->size - DIF_SEQUENCE_SIZE))
	{
	    if (size < 0)
		perror("ERROR: read");
	    else
		fputs("ERROR: Failed to read complete frame\n", stderr);
	    exit(1);
	}
	if (write(params->sock, buf, system->size) != (ssize_t)system->size)
	{
	    perror("ERROR: write");
	    exit(1);
	}

	frame_timestamp += frame_interval;
	frame_timer_wait(frame_timestamp);
    }
}

static int is_dv_file(int fd)
{
    uint8_t buf[DIF_SIGNATURE_SIZE];
    int is_dv;
    off_t orig = lseek(fd, 0, SEEK_CUR);

    /* Can't check a non-seekable file; assume it's valid */
    if (orig == -1)
	return 1;

    is_dv = (read(fd, buf, DIF_SIGNATURE_SIZE) == DIF_SIGNATURE_SIZE
	     && !memcmp(buf, DIF_SIGNATURE, DIF_SIGNATURE_SIZE));
    lseek(fd, orig, SEEK_SET);
    return is_dv;
}

int main(int argc, char ** argv)
{
    /* Initialise settings from configuration files. */
    dvswitch_read_config(handle_config);

    struct transfer_params params;
    params.opt_loop = false;

    /* Parse arguments. */

    int opt;
    while ((opt = getopt_long(argc, argv, "h:p:l", options, NULL)) != -1)
    {
	switch (opt)
	{
	case 'h':
	    free(mixer_host);
	    mixer_host = strdup(optarg);
	    break;
	case 'p':
	    free(mixer_port);
	    mixer_port = strdup(optarg);
	    break;
	case 'l':
	    params.opt_loop = true;
	    break;
	case 'H': /* --help */
	    usage(argv[0]);
	    return 0;
	default:
	    usage(argv[0]);
	    return 2;
	}
    }

    if (!mixer_host || !mixer_port)
    {
	fprintf(stderr, "%s: mixer hostname and port not defined\n",
		argv[0]);
	return 2;
    }

    if (optind != argc - 1)
    {
	if (optind == argc)
	{
	    fprintf(stderr, "%s: missing filename\n",
		    argv[0]);
	}
	else
	{
	    fprintf(stderr, "%s: excess argument \"%s\"\n",
		    argv[0], argv[optind + 1]);
	}
	usage(argv[0]);
	return 2;
    }

    const char * filename = argv[optind];

    /* Prepare to read the file and connect a socket to the mixer. */

    printf("INFO: Reading from %s\n", filename);
    params.file = open(filename, O_RDONLY, 0);
    if (params.file < 0)
    {
	perror("ERROR: open");
	return 1;
    }
    if (!is_dv_file(params.file)) {
        fprintf(stderr, "ERROR: %s is not a DV file\n", filename);
	return 1;
    }
    printf("INFO: Connecting to %s:%s\n", mixer_host, mixer_port);
    params.sock = create_connected_socket(mixer_host, mixer_port);
    assert(params.sock >= 0); /* create_connected_socket() should handle errors */
    if (write(params.sock, GREETING_SOURCE, GREETING_SIZE) != GREETING_SIZE)
    {
	perror("ERROR: write");
	exit(1);
    }

    transfer_frames(&params);

    close(params.sock);
    close(params.file);

    return 0;
}
